units/src/amount/result.rs

```diff
@@ -0,0 +1,573 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Provides a monodic result type that is used to return the result of
+//! doing mathematical operations (`core::ops`) on amount types.
+
+use core::{fmt, ops};
+
+use NumOpResult as R;
+
+use super::{Amount, SignedAmount};
+
+/// Result of an operation on [`Amount`] or [`SignedAmount`].
+///
+/// The type parameter `T` should be normally `Amout` or `SignedAmount`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[must_use]
+pub enum NumOpResult<T> {
+    /// Result of a successful mathematical operation.
+    Valid(T),
+    /// Result of an unsuccessful mathematical operation.
+    Error(NumOpError),
+}
+
+impl<T> NumOpResult<T> {
+    /// Returns the contained valid amount, consuming `self`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the result is an `Error`.
+    pub fn unwrap(self) -> T {
+        match self {
+            R::Valid(amount) => amount,
+            R::Error(_) => panic!("tried to unwrap an invalid result"),
```

Comment:
In 4d6911b12c0dec91ebf2ba1065224300e6231ad5:

Why not output the error as well? Also should change "result" to "numeric result" to give a bit of a hint.

Reply:
Oh, yes, we definitely should, `Result` does it too.

Reply:
Oh I didn't do that because `NumOpResult` is empty. I added it using `"{:?}"` - I"m not sure on the convention of using `Debug` vs `Dispayl` in `panic` messages.

Reply:
Oh, right, I'm dumb. We should probably keep it, since it's empty though we're risking forgetting about it when we add more stuff there.

> I"m not sure on the convention of using `Debug` vs `Dispayl` in `panic` messages.

`Result` uses `Debug` presumably because it's more widely applicable and panic messages are not meant for users anyway but here we have a fixed error type, so `Display` would've been equally fine. Currently it doesn't matter but later we should decide based on which gives the *developer* better information to aid with debugging.

Reply:
Ok, just to clarify for my education. A panic will be seen by a user but the message will be given to a developer so the message is tailored to the 'a developer has to debug this message'?

Reply:
In ideal world the user would never see panic or any output of `Debug` and only developers would see it. The point of `Debug` is that its formatting is so unimportant it can be auto-generated in vast majority of the cases. And `Debug` is used in `Result::unwrap` because it allows more types without confusing the users (because they shouldn't see it in the first place). But sometimes they have equal representation (numbers, strings...) I personally don't care here.